File: js/packages/openinference-core/src/trace/trace_config/OISpan.ts
import {
  Span,
  SpanContext,
  SpanStatus,
  TimeInput,
  Attributes,
  Exception,
  AttributeValue,
  Link,
} from "@opentelemetry/api";
import { TraceConfig } from "./types";
import { mask } from "./maskingRules";

/**
 * A wrapper around the OpenTelemetry {@link Span} interface that masks sensitive information based on the passed in {@link TraceConfig}.
 */
export class OISpan implements Span {
  private readonly span: Span;
  private readonly config: TraceConfig;

  constructor({ span, config }: { span: Span; config: TraceConfig }) {
    this.span = span;
    this.config = config;
  }

  setAttribute(key: string, value: AttributeValue): this {
    const maskedValue = mask({ config: this.config, key, value });
    if (maskedValue != null) {
      this.span.setAttribute(key, maskedValue);
    }
    return this;
  }

  setAttributes(attributes: Attributes): this {
    const maskedAttributes = Object.entries(attributes).reduce(
      (maskedAttributes, [key, value]) => {
        maskedAttributes[key] = mask({ config: this.config, key, value });
        return maskedAttributes;
      },
      {} as Attributes,
    );
    this.span.setAttributes(maskedAttributes);
    return this;
  }

  addEvent(
    name: string,
    attributesOrStartTime?: Attributes | TimeInput,
    startTime?: TimeInput,
  ): this {
    this.span.addEvent(name, attributesOrStartTime, startTime);
    return this;
  }

  setStatus(status: SpanStatus): this {
    this.span.setStatus(status);
    return this;
  }

  updateName(name: string): this {
    this.span.updateName(name);
    return this;
  }

  end(endTime?: TimeInput): void {
    this.span.end(endTime);
  }

  isRecording(): boolean {
    return this.span.isRecording();
  }

  recordException(exception: Exception, time?: TimeInput): void {
    this.span.recordException(exception, time);
  }

  spanContext(): SpanContext {
    return this.span.spanContext();
  }

  addLink(link: Link): this {
    this.span.addLink(link);
    return this;
  }

  addLinks(links: Link[]): this {
    this.span.addLinks(links);
    return this;
  }
}

File: java/openinference-instrumentation/src/main/java/com/arize/instrumentation/trace/config/OISpan.java
package com.arize.instrumentation.trace.config;

import com.arize.instrumentation.TraceConfig;
import io.opentelemetry.api.common.AttributeKey;
import io.opentelemetry.api.common.Attributes;
import io.opentelemetry.api.common.Value;
import io.opentelemetry.api.trace.Span;
import io.opentelemetry.api.trace.SpanContext;
import io.opentelemetry.api.trace.StatusCode;
import lombok.NonNull;
import lombok.RequiredArgsConstructor;

import java.util.concurrent.TimeUnit;

import static com.arize.instrumentation.trace.config.MaskingUtils.mask;

@RequiredArgsConstructor
public class OISpan implements Span {
    @NonNull
    private Span span;
    @NonNull
    private TraceConfig config;

    @Override
    public <T> Span setAttribute(AttributeKey<T> key, T value) {
        Value<T> maskedValue = mask(config, key.getKey(), (Value<T>) value);
        if (maskedValue != null) {
            span.setAttribute(key, maskedValue.getValue());
        }
        return span;
    }

    @Override
    public Span addEvent(String name, Attributes attributes) {
        span.addEvent(name, attributes);
        return span;
    }

    @Override
    public Span addEvent(String name, Attributes attributes, long timestamp, TimeUnit unit) {
        span.addEvent(name, attributes, timestamp, unit);
        return span;
    }

    @Override
    public Span setStatus(StatusCode statusCode, String description) {
        span.setStatus(statusCode, description);
        return span;
    }

    @Override
    public Span recordException(Throwable exception, Attributes additionalAttributes) {
        span.recordException(exception, additionalAttributes);
        return span;
    }

    @Override
    public Span updateName(String name) {
        span.updateName(name);
        return span;
    }

    @Override
    public void end() {
        span.end();
    }

    @Override
    public void end(long timestamp, TimeUnit unit) {
        span.end(timestamp, unit);
    }

    @Override
    public SpanContext getSpanContext() {
        return span.getSpanContext();
    }

    @Override
    public boolean isRecording() {
        return span.isRecording();
    }
}
